File: md5sum.py
import sys
import hashlib

def print_help():
    print("Usage: md5sum [options] [file]")
    print("Options:")
    print("  -b    Use binary mode")
    print("  -t    Use text mode (default)")
    print("  -h    Show this help")

def main():
    args = sys.argv[1:]
    
    if any(arg in ('-h', '--help', '/?') for arg in args):
        print_help()
        return 0
    
    binary_mode = '-b' in args
    text_mode = '-t' in args
    
    files = [arg for arg in args if not arg.startswith('-')]
    
    if not files:
        # Read from stdin
        data = sys.stdin.buffer.read() if binary_mode else sys.stdin.read()
        md5 = hashlib.md5(data).hexdigest()
        print(f"{md5}  -")
        return 0
    
    for file_path in files:
        try:
            mode = 'rb' if binary_mode else 'r'
            with open(file_path, mode) as f:
                data = f.read()
            md5 = hashlib.md5(data).hexdigest()
            print(f"{md5}  {file_path}")
        except FileNotFoundError:
            print(f"Error: File '{file_path}' not found")
        except Exception as e:
            print(f"Error: {e}")
    
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: md5sum.js
const crypto = require('crypto');
const fs = require('fs');

function print_help() {
    console.log("Usage: md5sum [options] [file]");
    console.log("Options:");
    console.log("  -b    Use binary mode");
    console.log("  -t    Use text mode (default)");
    console.log("  -h    Show this help");
}

const args = process.argv.slice(2);

if (args.includes('-h') || args.includes('--help')) {
    print_help();
    process.exit(0);
}

const binary_mode = args.includes('-b');
const files = args.filter(arg => !arg.startsWith('-'));

if (files.length === 0) {
    const chunks = [];
    process.stdin.on('readable', () => {
        let chunk;
        while ((chunk = process.stdin.read()) !== null) {
            chunks.push(chunk);
        }
    });
    process.stdin.on('end', () => {
        const data = Buffer.concat(chunks);
        const md5 = crypto.createHash('md5').update(data).digest('hex');
        console.log(`${md5}  -`);
    });
} else {
    files.forEach(file => {
        try {
            const buffer = fs.readFileSync(file, binary_mode ? null : 'utf8');
            const md5 = crypto.createHash('md5').update(buffer).digest('hex');
            console.log(`${md5}  ${file}`);
        } catch (error) {
            console.error(`Error: ${error.message}`);
        }
    });
}